### PR TITLE
Add ST_GeomFromWKB and ST_AsBinary functions

### PR DESCRIPTION
**Pull request by mpertierra**
Test Plan: Run `mvn -pl presto-geospatial -Dtest=TestGeoFunctions test`

### CODE REVIEW

File: presto-geospatial/src/test/java/com/facebook/presto/plugin/geospatial/TestGeoFunctions.java [hunk 1]
```diff
@@ -967,9 +967,10 @@ public void testSTGeometries()
 
     private void assertSTGeometries(String wkt, String... expected)
     {
-        assertFunction(String.format("transform(ST_Geometries(ST_GeometryFromText('%s')), x -> ST_ASText(x))", wkt), new ArrayType(VARCHAR), ImmutableList.copyOf(expected));
+        assertFunction(String.format("transform(ST_Geometries(ST_GeometryFromText('%s')), x -> ST_AsText(x))", wkt), new ArrayType(VARCHAR), ImmutableList.copyOf(expected));
```
Comment (martint) [inline presto-geospatial/src/test/java/com/facebook/presto/plugin/geospatial/TestGeoFunctions.java]: Unrelated change

File: presto-geospatial/src/test/java/com/facebook/presto/plugin/geospatial/TestGeoFunctions.java [hunk 2]
```diff
@@ -1008,4 +1010,24 @@ public void testSTGeometryType()
...
+        assertFunction(format("ST_AsText(ST_GeometryFromBinary(ST_AsBinary(ST_GeometryFromText('%s'))))", wkt), VARCHAR, expected);
+    }
+
+    @Test
+    public void testSTGeometryFromBinary()
+    {
+        assertToFromBinary("POINT EMPTY");
+        assertToFromBinary("LINESTRING (1 1, 2 2)");
+        assertToFromBinary("POLYGON ((1 1, 4 1, 4 4, 1 4, 1 1))");
+        assertToFromBinary("POLYGON ((1 1, 1 4, 4 4, 4 1))", "POLYGON ((1 1, 4 1, 4 4, 1 4, 1 1))");
```
Comment (martint) [inline presto-geospatial/src/test/java/com/facebook/presto/plugin/geospatial/TestGeoFunctions.java]: Why would these be different?

Comment (mpertierra) [reply presto-geospatial/src/test/java/com/facebook/presto/plugin/geospatial/TestGeoFunctions.java]: Apparently, this is expected behavior. See https://github.com/prestodb/presto/blob/master/presto-geospatial/src/test/java/com/facebook/presto/plugin/geospatial/TestGeoFunctions.java#L84.

File: presto-docs/src/main/sphinx/functions/geospatial.rst [hunk 3]
```diff
@@ -38,12 +38,20 @@ Constructors
...
+
 .. function:: ST_AsText(Geometry) -> varchar
 
     Returns the WKT representation of the geometry. For empty geometries,
     ``ST_AsText(ST_LineFromText('LINESTRING EMPTY'))`` will produce ``'MULTILINESTRING EMPTY'``
     and ``ST_AsText(ST_Polygon('POLYGON EMPTY'))`` will produce ``'MULTIPOLYGON EMPTY'``.
 
+.. function:: ST_AsBinary(Geometry) -> varbinary
+
+    Returns the WKB representation of the geometry.
```
Comment (mbasmanova) [inline presto-docs/src/main/sphinx/functions/geospatial.rst]: Would you specify what does this function return for an empty geometry?

File: presto-docs/src/main/sphinx/functions/geospatial.rst [hunk 4]
```diff
@@ -38,12 +38,20 @@ Constructors
 
     Returns a geometry type object from WKT representation.
 
+.. function:: ST_GeometryFromBinary(varbinary) -> Geometry
```
Comment (mbasmanova) [inline presto-docs/src/main/sphinx/functions/geospatial.rst]: Functions here are listed in alphabetical order. Could you move these new functions accordingly?

Also, the function name in the SQL/MM Part 3 specification is `ST_GeomFromWKB`, not `ST_GeometryFromWKB`. Could you please update? (I don't know why we have `ST_GeometryFromText` and not `ST_GeomFromText`. This needs to change as well at some point.)

screen shot 2018-09-04 at 10 43 32 am

File: presto-geospatial/src/main/java/com/facebook/presto/plugin/geospatial/GeoFunctions.java [hunk 5]
```diff
@@ -169,6 +170,14 @@ public static Slice stGeometryFromText(@SqlType(StandardTypes.VARCHAR) Slice inp
         return serialize(geometryFromText(input));
     }
 
+    @Description("Returns a Geometry type object from Well-Known Text representation (WKT)")
```
Comment (mbasmanova) [inline presto-geospatial/src/main/java/com/facebook/presto/plugin/geospatial/GeoFunctions.java]: Copy-paste error? This should refer to WKB, not WKT.

File: presto-geospatial/src/main/java/com/facebook/presto/plugin/geospatial/GeoFunctions.java [hunk 6]
```diff
@@ -1044,6 +1061,19 @@ private static OGCGeometry geometryFromText(Slice input)
...
     }
 
+    private static OGCGeometry geometryFromBinary(Slice input)
+    {
+        OGCGeometry geometry;
+        try {
+            geometry = OGCGeometry.fromBinary(input.toByteBuffer());
+        }
+        catch (IllegalArgumentException e) {
+            throw new PrestoException(INVALID_FUNCTION_ARGUMENT, "Invalid WKB: " + input.toByteBuffer(), e);
```
Comment (mbasmanova) [inline presto-geospatial/src/main/java/com/facebook/presto/plugin/geospatial/GeoFunctions.java]: Including the binary input in the error message is probably not useful, is it? Perhaps, include `e.getMessage()` instead.

File: presto-geospatial/src/test/java/com/facebook/presto/plugin/geospatial/TestGeoFunctions.java (hunk 1 above)
Comment (mbasmanova) [reply presto-geospatial/src/test/java/com/facebook/presto/plugin/geospatial/TestGeoFunctions.java]: +1

File: presto-geospatial/src/test/java/com/facebook/presto/plugin/geospatial/TestGeoFunctions.java [hunk 7]
```diff
@@ -967,9 +967,10 @@ public void testSTGeometries()
 
     private void assertSTGeometries(String wkt, String... expected)
     {
-        assertFunction(String.format("transform(ST_Geometries(ST_GeometryFromText('%s')), x -> ST_ASText(x))", wkt), new ArrayType(VARCHAR), ImmutableList.copyOf(expected));
+        assertFunction(String.format("transform(ST_Geometries(ST_GeometryFromText('%s')), x -> ST_AsText(x))", wkt), new ArrayType(VARCHAR), ImmutableList.copyOf(expected));
     }
 
+    @Test
```
[the remark this answers is not in the document]
Comment (mbasmanova) [reply presto-geospatial/src/test/java/com/facebook/presto/plugin/geospatial/TestGeoFunctions.java]: @mpertierra Marcos, would you move this change into a separate commit?

File: presto-geospatial/src/test/java/com/facebook/presto/plugin/geospatial/TestGeoFunctions.java (hunk 2 above)
Comment (mbasmanova) [reply presto-geospatial/src/test/java/com/facebook/presto/plugin/geospatial/TestGeoFunctions.java]: @mpertierra Marcos, in the canonical representation the vertexes are listed in counterclockwise order. In this test case, the vertexes are listed in clockwise order. `assertToFromBinary("POLYGON ((1 1, 4 1, 4 4, 1 4, 1 1))"` should work.

The other test needs to be modified as well. If you could do that in a separate commit, that would be appreciated.

File: presto-geospatial/src/test/java/com/facebook/presto/plugin/geospatial/TestGeoFunctions.java [hunk 8]
```diff
@@ -1008,4 +1010,24 @@ public void testSTGeometryType()
...
+
+    private void assertToFromBinary(String wkt, String expected)
+    {
+        assertFunction(format("ST_AsText(ST_GeometryFromBinary(ST_AsBinary(ST_GeometryFromText('%s'))))", wkt), VARCHAR, expected);
+    }
+
+    @Test
+    public void testSTGeometryFromBinary()
+    {
+        assertToFromBinary("POINT EMPTY");
```
Comment (mbasmanova) [inline presto-geospatial/src/test/java/com/facebook/presto/plugin/geospatial/TestGeoFunctions.java]: For consistency, let's rename this function to `assertGeomFromBinary` (assuming the test will be renamed to `testSTGeomFromBinary`).

File: presto-geospatial/src/test/java/com/facebook/presto/plugin/geospatial/TestGeoFunctions.java [hunk 9]
```diff
@@ -1008,4 +1010,24 @@ public void testSTGeometryType()
...
+        assertToFromBinary(wkt, wkt);
+    }
+
+    private void assertToFromBinary(String wkt, String expected)
+    {
+        assertFunction(format("ST_AsText(ST_GeometryFromBinary(ST_AsBinary(ST_GeometryFromText('%s'))))", wkt), VARCHAR, expected);
+    }
+
+    @Test
+    public void testSTGeometryFromBinary()
```
Comment (mbasmanova) [inline presto-geospatial/src/test/java/com/facebook/presto/plugin/geospatial/TestGeoFunctions.java]: Could you add tests for multi-geometries and geometry collections as well tests for invalid binaries?

File: presto-docs/src/main/sphinx/functions/geospatial.rst (hunk 4 above)
Comment (mpertierra) [reply presto-docs/src/main/sphinx/functions/geospatial.rst]: @mbasmanova The functions in this section are not in alphabetical order, so I can not put the new functions in that way. Should I reorder the functions in this section or would that have to be in a separate PR?

Comment (mbasmanova) [reply presto-docs/src/main/sphinx/functions/geospatial.rst]: @mpertierra Oh... Thanks for pointing it out. Would you fix the order of existing functions in a separate commit? All commits can be part of this PR. No need to create a new PR.

File: presto-geospatial/src/main/java/com/facebook/presto/plugin/geospatial/GeoFunctions.java [hunk 10]
```diff
@@ -1044,6 +1061,19 @@ private static OGCGeometry geometryFromText(Slice input)
         return geometry;
     }
 
+    private static OGCGeometry geometryFromBinary(Slice input)
+    {
+        OGCGeometry geometry;
```
Comment (nezihyigitbasi) [inline presto-geospatial/src/main/java/com/facebook/presto/plugin/geospatial/GeoFunctions.java]: Add `requireNonNull(input, "input is null")`

File: presto-docs/src/main/sphinx/functions/geospatial.rst (hunk 3 above)
Comment (mpertierra) [reply presto-docs/src/main/sphinx/functions/geospatial.rst]: @mbasmanova For empty geometries,
    ``ST_AsBinary(ST_LineFromText('LINESTRING EMPTY'))`` will produce ``[01 02 00 00 00 00 00 00   00]``
    and ``ST_AsBinary(ST_Polygon('POLYGON EMPTY'))`` will produce ``[01 03 00 00 00 00 00 00   00]``.  Is this what we want in the docs?

Comment (mbasmanova) [reply presto-docs/src/main/sphinx/functions/geospatial.rst]: @mpertierra I see. Somehow I was thinking that it might produce `null`. Looks like it will actually preserve the type of the geometry which is nice. No, in this case, I think it is fine to skip this detail.

File: presto-geospatial/src/test/java/com/facebook/presto/plugin/geospatial/TestGeoFunctions.java [hunk 11]
```diff
@@ -1010,4 +1010,40 @@ public void testSTGeometryType()
...
+        // valid nonempty geometries
+        assertGeomFromBinary("POINT (1 2)");
+        assertGeomFromBinary("MULTIPOINT ((1 2), (3 4))");
+        assertGeomFromBinary("LINESTRING (0 0, 1 2, 3 4)");
+        assertGeomFromBinary("MULTILINESTRING ((1 1, 5 1), (2 4, 4 4))");
+        assertGeomFromBinary("POLYGON ((0 0, 1 0, 1 1, 0 1, 0 0))");
+        assertGeomFromBinary("MULTIPOLYGON (((1 1, 3 1, 3 3, 1 3, 1 1)), ((2 4, 6 4, 6 6, 2 6, 2 4)))");
+        assertGeomFromBinary("GEOMETRYCOLLECTION (POINT (1 2), LINESTRING (0 0, 1 2, 3 4), POLYGON ((0 0, 1 0, 1 1, 0 1, 0 0)))");
+
+        // invalid geometries
```
[the remark this answers is not in the document]
Comment (mpertierra) [reply presto-geospatial/src/test/java/com/facebook/presto/plugin/geospatial/TestGeoFunctions.java]: It's invalid because `ST_IsValid` will return false for `LINESTRING (0 0, 0 1, 0 1, 1 1, 1 0, 0 0)`.

File: presto-geospatial/src/main/java/com/facebook/presto/plugin/geospatial/GeoFunctions.java [hunk 12]
```diff
@@ -1044,6 +1062,20 @@ private static OGCGeometry geometryFromText(Slice input)
...
 
+    private static OGCGeometry geomFromBinary(Slice input)
+    {
+        requireNonNull(input, "input is null");
+        OGCGeometry geometry;
+        try {
+            geometry = OGCGeometry.fromBinary(input.toByteBuffer());
+        }
+        catch (IllegalArgumentException | IndexOutOfBoundsException e) {
+            throw new PrestoException(INVALID_FUNCTION_ARGUMENT, "Invalid WKB: " + e.toString(), e);
```
[the remark this answers is not in the document]
Comment (mpertierra) [reply presto-geospatial/src/main/java/com/facebook/presto/plugin/geospatial/GeoFunctions.java]: @mbasmanova `getMessage()` for `java.lang.IndexOutOfBoundsException` here returns `null`

[the remark this answers is not in the document]
Comment (mbasmanova) [reply presto-geospatial/src/main/java/com/facebook/presto/plugin/geospatial/GeoFunctions.java]: @mpertierra That's not great, but I think it is better than `java.lang.IndexOutOfBoundsException`.

[the remark this answers is not in the document]
Comment (mpertierra) [reply presto-geospatial/src/main/java/com/facebook/presto/plugin/geospatial/GeoFunctions.java]: @mbasmanova Ok, I will make the change

[the remark this answers is not in the document]
Comment (mbasmanova) [reply presto-geospatial/src/main/java/com/facebook/presto/plugin/geospatial/GeoFunctions.java]: @mpertierra Perhaps, the error message should be just "Invalid WKB"?

[the remark this answers is not in the document]
Comment (mpertierra) [reply presto-geospatial/src/main/java/com/facebook/presto/plugin/geospatial/GeoFunctions.java]: @mbasmanova Done!
